File: src/main/dnd_combat/combat_menu/popup/action_panel/DamageAmountPanel.java
package combat_menu.popup.action_panel;

import __main.CombatMain;
import character_info.combatant.Combatant;
import combat_menu.listener.DieRollListener;
import combat_menu.listener.IntegerFieldListener;
import damage_implements.Effect;
import damage_implements.Implement;
import damage_implements.Spell;
import damage_implements.Weapon;
import format.ColorStyle;
import format.SwingStyles;
import util.Message;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import java.awt.*;

import static util.Message.informIllusion;

public class DamageAmountPanel extends JPanel {

    private final ActionPanel root;

    private final Combatant target;
    private final Implement implement;
    private final Combatant attacker;

    private final JTextField mainDamageField = new JTextField(8);
    private final JCheckBox otherBonusDamageCheck = new JCheckBox("Other Bonus Damage?");
    private final JTextField otherBonusDamageField = new JTextField(8);
    private final JLabel summaryLabel = new JLabel("Total Damage: 0");
    private final JButton okButton = new JButton("Deal Damage");

    public static DamageAmountPanel newInstance(Implement implement, Combatant target, ActionPanel root) {
        return new DamageAmountPanel(implement, target, root);
    }

    private DamageAmountPanel(Implement implement, Combatant target, ActionPanel root) {
        this.root = root;

        this.implement = implement;
        this.target = target;
        this.attacker = CombatMain.QUEUE.getCurrentCombatant();

        setBorder(BorderFactory.createEmptyBorder(15, 15, 15, 15));
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));

        JPanel damagePanel = getDamagePanel();
        JPanel bonusPanel = getBonusPanel();
        JPanel summaryPanel = getSummaryPanel();

        okButton.setBackground(ColorStyle.DARKER_RED.getColor());
        JPanel okCancelPanel = SwingStyles.getConfirmCancelPanel(
                okButton,
                e -> logAndFinish(),
                e -> root.returnToButtons()
        );

        add(damagePanel);
        add(Box.createVerticalStrut(10));
        add(bonusPanel);
        add(Box.createVerticalStrut(10));
        add(summaryPanel);
        add(Box.createVerticalStrut(15));
        add(okCancelPanel);

        setupListeners();
        updateUIState();

        SwingUtilities.invokeLater(mainDamageField::requestFocusInWindow);
    }

    private JPanel getDamagePanel() {
        JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        SwingStyles.addLabeledBorder(panel, "Main Damage");

        JLabel label = new JLabel("Enter " + implement.damageString() + ":");

        mainDamageField.setFont(mainDamageField.getFont().deriveFont(16f));
        mainDamageField.setHorizontalAlignment(JTextField.CENTER);

        panel.add(label);
        panel.add(mainDamageField);

        return panel;
    }

    private JPanel getBonusPanel() {
        JPanel panel = new JPanel();
        panel.setLayout(new FlowLayout(FlowLayout.LEFT));
        SwingStyles.addLabeledBorder(panel, "Bonus Damage");

        otherBonusDamageField.setEnabled(false);

        panel.add(otherBonusDamageCheck);
        panel.add(otherBonusDamageField);

        return panel;
    }

    private JPanel getSummaryPanel() {
        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
        SwingStyles.addLabeledBorder(panel, "Breakdown");

        if (implement instanceof Weapon && !implement.isManual()) {
            panel.add(new JLabel("Stat Bonus: +" + attacker.mod(implement.stat())));
        }
        if (implement.isHalfDamage()) {
            panel.add(new JLabel("Damage will be halved (Failed Attack)"));
        }
        if (target.isHexedBy(attacker)) {
            panel.add(new JLabel("Hex: include +1d6 in bonus field"));
        }

        summaryLabel.setFont(summaryLabel.getFont().deriveFont(Font.BOLD, 16f));
        summaryLabel.setForeground(ColorStyle.DARKER_RED.getColor());
        summaryLabel.setAlignmentX(Component.CENTER_ALIGNMENT);

        panel.add(Box.createVerticalStrut(5));
        panel.add(summaryLabel);

        return panel;
    }

    private void setupListeners() {
        DocumentListener updateListener = new DocumentListener() {
            public void insertUpdate(DocumentEvent e) {
                updateUIState();
            }

            public void removeUpdate(DocumentEvent e) {
                updateUIState();
            }

            public void changedUpdate(DocumentEvent e) {
                updateUIState();
            }
        };

        mainDamageField.getDocument().addDocumentListener(updateListener);
        otherBonusDamageField.getDocument().addDocumentListener(updateListener);

        if (implement.isManual()) {
            mainDamageField.addKeyListener(new IntegerFieldListener());
        } else {
            mainDamageField.addKeyListener(
                    new DieRollListener(
                            implement.numDice(),
                            implement.dieSize(),
                            mainDamageField
                    )
            );
        }

        otherBonusDamageField.addKeyListener(new IntegerFieldListener());
        otherBonusDamageCheck.addActionListener(e -> {
            boolean enabled = otherBonusDamageCheck.isSelected();
            otherBonusDamageField.setEnabled(enabled);

            if (!enabled) {
                otherBonusDamageField.setText("");
            }

            updateUIState();
        });

        okButton.addActionListener(e -> logAndFinish());
    }

    private int calculateTotal() {
        int damage = getMainDamage() + getBonusDamage();

        if (implement.isManual()) {
            return damage;
        }
        if (implement.isHalfDamage()) {
            damage /= 2;
        }
        if (implement instanceof Weapon) {
            damage += attacker.mod(implement.stat());
        }

        return Math.max(0, damage);
    }

    private int getMainDamage() {
        String text = mainDamageField.getText().trim();
        if (text.isEmpty()) return 0;
        return Integer.parseInt(text);
    }

    private int getBonusDamage() {
        if (!otherBonusDamageCheck.isSelected()) return 0;

        String text = otherBonusDamageField.getText().trim();
        if (text.isEmpty()) return 0;

        return Integer.parseInt(text);
    }

    private void updateUIState() {
        okButton.setBackground(ColorStyle.DARKER_RED.getColor());
        try {
            int total = calculateTotal();
            if (total > 0) {
                okButton.setEnabled(true);
                okButton.setText("Deal " + total + " Damage");
                summaryLabel.setText("Total Damage: " + total);
            } else {
                okButton.setEnabled(false);
                okButton.setText("Enter Damage");
                summaryLabel.setText("Total Damage: 0");
            }
        } catch (NumberFormatException e) {
            okButton.setEnabled(false);
            okButton.setText("Numbers Only");
        }
    }

    private void logAndFinish() {
        int total = calculateTotal();
        target.damage(total);

        if (implement instanceof Spell spell) {
            Effect effect = spell.effect();
            attacker.putEffect(target, effect);

            if (effect.equals(Effect.ILLUSION)) {
                informIllusion(target);
            } else if (effect.equals(Effect.ADVANTAGE_SOON)) {
                Message.template("Advantage on next attack against " + target + " this turn.");
            }
        }

        CombatMain.COMBAT_MENU.update();
        CombatMain.checkWinConditions();

        root.returnToButtons();
    }
}

File: src/main/dnd_combat/combat_menu/action_panel/damage_panel/DamageAmountPanel.kt
package combat_menu.action_panel.damage_panel

import __main.EncounterInfo
import __main.Main
import character_info.combatant.Combatant
import combat_menu.action_panel.ActionPanel
import combat_menu.listener.DieRollListener
import combat_menu.listener.IntegerFieldListener
import damage_implements.Effect
import damage_implements.Implement
import damage_implements.Spell
import damage_implements.Weapon
import format.ColorStyle
import format.SwingStyles
import util.Message
import java.awt.FlowLayout
import java.awt.Font
import java.awt.event.ActionEvent
import javax.swing.*
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener
import kotlin.math.max

class DamageAmountPanel private constructor(
    private val implement: Implement,
    private val target: Combatant,
    private val root: ActionPanel
) : JPanel() {

    private val attacker: Combatant = EncounterInfo.getCurrentCombatant()

    private val mainDamageField = JTextField(8)
    private val otherBonusDamageCheck = JCheckBox("Other Bonus Damage?")
    private val otherBonusDamageField = JTextField(8)
    private val summaryLabel = JLabel("Total Damage: 0")
    private val okButton = JButton("Deal Damage")

    init {
        border = BorderFactory.createEmptyBorder(15, 15, 15, 15)
        layout = BoxLayout(this, BoxLayout.Y_AXIS)

        val damagePanel = damagePanel
        val bonusPanel = bonusPanel
        val summaryPanel = summaryPanel

        okButton.background = ColorStyle.DARKER_RED.color
        val okCancelPanel = SwingStyles.getConfirmCancelPanel(
            okButton,
            { _: ActionEvent? -> logAndFinish() },
            { _: ActionEvent? -> root.returnToButtons() }
        )

        SwingStyles.addComponents(
            this,
            damagePanel, Box.createVerticalStrut(10),
            bonusPanel, Box.createVerticalStrut(10),
            summaryPanel, Box.createVerticalStrut(15),
            okCancelPanel
        )

        setupListeners()
        updateUIState()

        SwingUtilities.invokeLater { mainDamageField.requestFocusInWindow() }
    }

    private val damagePanel: JPanel
        get() {
            val panel = JPanel(FlowLayout(FlowLayout.LEFT))
            SwingStyles.addLabeledBorder(panel, "Main Damage")

            val label = JLabel("Enter " + implement.damageString() + ":")

            mainDamageField.font = mainDamageField.font.deriveFont(16f)
            mainDamageField.horizontalAlignment = JTextField.CENTER

            panel.add(label)
            panel.add(mainDamageField)

            return panel
        }

    private val bonusPanel: JPanel
        get() {
            val panel = JPanel()
            panel.layout = FlowLayout(FlowLayout.LEFT)
            SwingStyles.addLabeledBorder(panel, "Bonus Damage")

            otherBonusDamageField.isEnabled = false

            panel.add(otherBonusDamageCheck)
            panel.add(otherBonusDamageField)

            return panel
        }

    private val summaryPanel: JPanel
        get() {
            val panel = JPanel()
            panel.layout = BoxLayout(panel, BoxLayout.Y_AXIS)
            SwingStyles.addLabeledBorder(panel, "Breakdown")

            if (implement is Weapon && !implement.isManual()) {
                panel.add(JLabel("Stat Bonus: +" + attacker.mod(implement.stat())))
            }
            if (implement.isHalfDamage) {
                panel.add(JLabel("Damage will be halved (Failed Attack)"))
            }
            if (target.isHexedBy(attacker)) {
                panel.add(JLabel("Hex: include +1d6 in bonus field"))
            }

            summaryLabel.font = summaryLabel.font.deriveFont(Font.BOLD, 16f)
            summaryLabel.foreground = ColorStyle.DARKER_RED.color
            summaryLabel.alignmentX = CENTER_ALIGNMENT

            panel.add(Box.createVerticalStrut(5))
            panel.add(summaryLabel)

            return panel
        }

    private fun setupListeners() {
        val updateListener: DocumentListener = object : DocumentListener {
            override fun insertUpdate(e: DocumentEvent) {
                updateUIState()
            }

            override fun removeUpdate(e: DocumentEvent) {
                updateUIState()
            }

            override fun changedUpdate(e: DocumentEvent) {
                updateUIState()
            }
        }

        mainDamageField.document.addDocumentListener(updateListener)
        otherBonusDamageField.document.addDocumentListener(updateListener)

        if (implement.isManual) {
            mainDamageField.addKeyListener(IntegerFieldListener())
        } else {
            mainDamageField.addKeyListener(
                DieRollListener(
                    implement.numDice(),
                    implement.dieSize(),
                    mainDamageField
                )
            )
        }

        otherBonusDamageField.addKeyListener(IntegerFieldListener())
        otherBonusDamageCheck.addActionListener {
            val enabled = otherBonusDamageCheck.isSelected
            otherBonusDamageField.isEnabled = enabled

            if (!enabled) {
                otherBonusDamageField.text = ""
            }
            updateUIState()
        }
    }

    private fun calculateTotal(): Int {
        var damage = mainDamage + bonusDamage

        if (implement.isManual) {
            return damage
        }
        if (implement.isHalfDamage) {
            damage /= 2
        }
        if (implement is Weapon) {
            damage += attacker.mod(implement.stat())
        }

        return max(0.0, damage.toDouble()).toInt()
    }

    private val mainDamage: Int
        get() {
            val text = mainDamageField.text.trim { it <= ' ' }
            if (text.isEmpty()) return 0
            return text.toInt()
        }

    private val bonusDamage: Int
        get() {
            if (!otherBonusDamageCheck.isSelected) return 0

            val text = otherBonusDamageField.text.trim { it <= ' ' }
            if (text.isEmpty()) return 0

            return text.toInt()
        }

    private fun updateUIState() {
        okButton.background = ColorStyle.DARKER_RED.color
        try {
            val total = calculateTotal()
            if (total > 0) {
                okButton.isEnabled = true
                okButton.text = "Deal $total Damage"
                summaryLabel.text = "Total Damage: $total"
            } else {
                okButton.isEnabled = false
                okButton.text = "Enter Damage"
                summaryLabel.text = "Total Damage: 0"
            }
        } catch (e: NumberFormatException) {
            okButton.isEnabled = false
            okButton.text = "Numbers Only"
        }
    }

    private fun logAndFinish() {
        val total = calculateTotal()
        target.damage(total)

        if (implement is Spell) {
            val effect = implement.effect()
            attacker.putEffect(target, effect)

            if (effect == Effect.ILLUSION) {
                Message.informIllusion(target)
            } else if (effect == Effect.ADVANTAGE_SOON) {
                Message.template("Advantage on next attack against $target this turn.")
            }
        }

        Main.logAction()

        root.returnToButtons()
    }

    companion object {
        @JvmStatic
        fun newInstance(implement: Implement, target: Combatant, root: ActionPanel): DamageAmountPanel {
            return DamageAmountPanel(implement, target, root)
        }
    }
}